keps/sig-storage/0037-20190129-PVCTaint.md

```diff
@@ -0,0 +1,222 @@
+  - sig-architecture
+reviewers:
+  - "@saad-ali"
+  - "@liggitt"
+  - "@davidopp"
+approvers:
+  - "@liggitt"
+  - "@saad-ali"
+editor: TBD
+creation-date: 2019-1-29
+last-updated: 2019-1-29
+status: implementable
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+PVC/PV Taint and Toleration
+
+## Table of Contents
+
+  * [Title](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Workarounds](#workarounds)
+      * [Alternatives](#alternatives)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+The current Taint and Toleration feature allows users to mark a node (taint the node) so that no pods can be scheduled to it, unless a pod explicitly tolerates the taint. This proposal is trying to extend this concept on a volume. User or admin (or controller) can mark a volume (taint the PVC or PV) so that no pods could start to use them unless a pod can tolerate the taint. This feature is very useful for some storage operations such as volume snapshot/clone and data population.  
+
+## Motivation
+
+Currently in our volume management design, k8s only supports provisioning an empty volume and a volume can be used by a pod immediately after it is created. With the new features on storage and data management becoming available such as snapshots, clone, and data population, data might be restored or populated to the volumes. We need a way to indicate when the data is available on the volume so that pod can start using it. 
+
+Previously, we proposed to use Readiness Gate for this problem, however, this approach lacks the support the scenarios that some special Pods (e.g., Pods for populating the data) can be allowed to use the volume even when the data is not ready.
+
+Taint and toleration feature currently used for marking node and reacting based
+on the taint effect seems also provides a nice way to solve this volume issue.
+
+### Goals
+
+This proposal is trying to investigate the taint and toleration feature applied
+to PVC/PV. It will identify the use cases that this feature could be applied to
+and describe the workflow of how this feature can be used.
+
+### Non-Goals
+
+
+## Proposal
+
+The idea of Taints and tolerations is that they work together to ensure that pods are not scheduled or executed onto inappropriate nodes. When a taint is applied to a node, this marks that the node should not accept (or execute based on the TaintEffect) any pods that do not tolerate the taints. Tolerations are applied to pods, and allow (but do not require) the pods to schedule onto nodes with matching taints.
+
+Currently, Kubernetes already supports three types of taint effects: 
```

Comment:
This paragraph still implies that the existing taint effects are applicable to VolumeTaint.
If you want to keep it, please replace the first sentence with something like: "Kubernetes node taints support three effects:"

Reply:
Just echoing the note from @bsalamat , that section confuses things a bit, even if we for some reason want node based taints/toleartions that needs to be explicit, the key use case I think we need to define in terms of the kep is capabilities for pod based taints